### PR TITLE
doc(ADR): only public types in public APIs

### PR DESCRIPTION
**Pull request by coryan**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8042)

### CODE REVIEW

File: doc/adr/2022-01-21-only-public-types-in-public-APIs.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+**Title**: use only public types in public APIs, aliases are no exceptions
+
+**Status**: proposed <!-- TODO(coryan) - change to accepted before merge -->
+
+**Context**: We want our public APIs to be fully documented, easy to discover,
+and *stable*.  We need the flexibility to change the implementation details of
+our public APIs.  We label symbols that are subject to change by placing them
+in a `::internal::` or `::${library}_internal::` namespace.  This is all well
```
Comment (devbww) [inline doc/adr/2022-01-21-only-public-types-in-public-APIs.md:8]: s/in a/in an/

File: doc/adr/2022-01-21-only-public-types-in-public-APIs.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
+**Title**: use only public types in public APIs, aliases are no exceptions
```
Comment (devbww) [inline doc/adr/2022-01-21-only-public-types-in-public-APIs.md:1]: s/,/, and/ perhaps?

File: doc/adr/2022-01-21-only-public-types-in-public-APIs.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+**Status**: proposed <!-- TODO(coryan) - change to accepted before merge -->
+
+**Context**: We want our public APIs to be fully documented, easy to discover,
+and *stable*.  We need the flexibility to change the implementation details of
+our public APIs.  We label symbols that are subject to change by placing them
+in a `::internal::` or `::${library}_internal::` namespace.  This is all well
+and good, but sometimes we have returned aliases to these internal symbols in
```
Comment (devbww) [inline doc/adr/2022-01-21-only-public-types-in-public-APIs.md:9]: s/returned/used/

File: doc/adr/2022-01-21-only-public-types-in-public-APIs.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+our public APIs.  We should not do this, it leaves important parts of the public
+API undocumented, and makes it unclear if the API is stable.
+
+**Decision**: never require (take as an argument) or return an internal type in
```
Comment (devbww) [inline doc/adr/2022-01-21-only-public-types-in-public-APIs.md:13]: s/never/Never/

File: doc/adr/2022-01-21-only-public-types-in-public-APIs.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+the public APIs.  If a type is part of the public API, then the type needs to be
+moved out of the internal namespace(s).  If possible, just expose an interface
```
Comment (devbww) [inline doc/adr/2022-01-21-only-public-types-in-public-APIs.md:15]: s/(s)//?

File: doc/adr/2022-01-21-only-public-types-in-public-APIs.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+in the public API, and keep the implementation details in the internal
+namespace.
+
+**Consequences**: on the positive side, our public APIs become better
```
Comment (devbww) [inline doc/adr/2022-01-21-only-public-types-in-public-APIs.md:19]: s/on/On/

**Review by devbww — COMMENTED** [review]
All nits.

File: doc/adr/2022-01-21-only-public-types-in-public-APIs.md (hunk 1 above)
Comment (coryan) [reply doc/adr/2022-01-21-only-public-types-in-public-APIs.md]: Fixed.

File: doc/adr/2022-01-21-only-public-types-in-public-APIs.md (hunk 2 above)
Comment (coryan) [reply doc/adr/2022-01-21-only-public-types-in-public-APIs.md]: Fixed.

File: doc/adr/2022-01-21-only-public-types-in-public-APIs.md (hunk 3 above)
Comment (coryan) [reply doc/adr/2022-01-21-only-public-types-in-public-APIs.md]: Fixed.

File: doc/adr/2022-01-21-only-public-types-in-public-APIs.md (hunk 4 above)
Comment (coryan) [reply doc/adr/2022-01-21-only-public-types-in-public-APIs.md]: Fixed.

File: doc/adr/2022-01-21-only-public-types-in-public-APIs.md (hunk 5 above)
Comment (coryan) [reply doc/adr/2022-01-21-only-public-types-in-public-APIs.md]: Fixed.

File: doc/adr/2022-01-21-only-public-types-in-public-APIs.md (hunk 6 above)
Comment (coryan) [reply doc/adr/2022-01-21-only-public-types-in-public-APIs.md]: Fixed.

**Comment by coryan** [comment]
@devjgm @dbolduc @devbww @scotthart for this I would like a LGTM from each of you.  Thanks.

**Review by devjgm — APPROVED** [review]

**Review by devbww — APPROVED** [review]

**Review by dbolduc — APPROVED** [review]
I guess we are doing double spaces after periods now

**Review by scotthart — APPROVED** [review]

**Pull request merged**